src/main/java/io/stargate/sgv2/jsonapi/service/operation/filters/table/codecs/MapCodecs.java

```diff
@@ -59,55 +86,96 @@ private static Map<String, Object> toCqlMap(
+  /**
+   * Method that will convert from driver-provided CQL Map type into JSON output.
+   *
+   * <p>On map read path, it wil in object format when map column is on text/ascii keys, return in
+   * tuple format for other key types. E.G. <ui>
+   * <li>Map<Double,Double> cql->json [[1.0, 1.0], [2.0, 2.0]]
+   * <li>Map<UUID, TEXT> cql-json [["d3b07384-d113-4c5a-9d3d-92b3a4d1f5a3", "apple"],
+   *     [[d3b07384-d113-4c5a-9d3d-92b3a4d1f5a4", "banana"]]
+   * <li>Map<TEXT, TEXT> cql-json {"apple": "apple", "banana": "banana"}
+   * <li>Map<ASCII, ASCII> cql-json {"apple": "apple", "banana": "banana"} </ui>
+   *
+   * @param keyType Cql map key type
+   * @param keyCodec0 Codec that can handle the keys of the map
+   * @param valueCodec0 Codec that can handle the values of the map
+   * @param objectMapper Jackson object mapper
+   * @param mapValue Cql map value
+   * @return JsonNode representation of the Cql map
+   */
   private static JsonNode cqlMapToJsonNode(
-      JSONCodec<?, ?> valueCodec0, ObjectMapper objectMapper, Object mapValue)
+      DataType keyType,
+      JSONCodec<?, ?> keyCodec0,
+      JSONCodec<?, ?> valueCodec0,
+      ObjectMapper objectMapper,
+      Object mapValue)
       throws ToJSONCodecException {
+    JSONCodec<?, Object> keyCodec = (JSONCodec<?, Object>) keyCodec0;
     JSONCodec<?, Object> valueCodec = (JSONCodec<?, Object>) valueCodec0;
-    final ObjectNode result = objectMapper.createObjectNode();
-    for (Map.Entry<Object, Object> entry : ((Map<Object, Object>) mapValue).entrySet()) {
-      Object key = entry.getKey();
-      if (key instanceof String strKey) {
+
+    if (keyType.equals(DataTypes.TEXT) || keyType.equals(DataTypes.ASCII)) {
+      final ObjectNode result = objectMapper.createObjectNode();
+      for (Map.Entry<String, Object> entry : ((Map<String, Object>) mapValue).entrySet()) {
+        String key = entry.getKey();
         Object value = entry.getValue();
         if (value == null) {
-          result.putNull(strKey);
+          result.putNull(key);
         } else {
-          result.put(strKey, valueCodec.toJSON(objectMapper, value));
+          result.put(key, valueCodec.toJSON(objectMapper, value));
         }
-      } else {
-        throw new ToJSONCodecException(
-            mapValue,
-            valueCodec.targetCQLType(),
-            String.format(
-                "expected String key, got: (%s) %s",
-                Optional.ofNullable(key).map(Object::getClass).map(Class::getName).orElse("null"),
-                String.valueOf(key)));
       }
+      return result;
+    } else {
+      final ArrayNode result = objectMapper.createArrayNode();
+      for (Map.Entry<Object, Object> entry : ((Map<Object, Object>) mapValue).entrySet()) {
+        var tupleEntry =
+            objectMapper
+                .createArrayNode()
+                .add(keyCodec.toJSON(objectMapper, entry.getKey()))
+                .add(
+                    entry.getValue() == null
```

Comment:
not: make the value access the same for both cases , and use addNull() on the array node to be explicit